### PR TITLE
Fix Redis command stats test coverage by exercising paths

### PR DESCRIPTION
**Pull request by FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg**
Exercises the code path used by the optional (off by default) RedisCommandStats. Assuming this increases coverage, I'll make it reach the coverage threshold and will ask Henry to review.

### CODE REVIEW

File: test/extensions/filters/network/common/redis/client_impl_test.cc [hunk 1]
```diff
@@ -281,10 +281,24 @@ TEST_F(RedisClientImplTest, BatchWithTimerCancelledByBufferFlush) {
...
+  ReadPolicy readPolicy() const override { return ReadPolicy::Master; }
+  uint32_t maxUpstreamUnknownConnections() const override { return 0; }
+  bool enableCommandStats() const override { return true; }
+};
+
 TEST_F(RedisClientImplTest, Basic) {
   InSequence s;
 
-  setup();
+  setup(std::make_unique<ConfigEnableCommandStats>());
```
Comment (mattklein123) [inline test/extensions/filters/network/common/redis/client_impl_test.cc]: Can you actually test the stats are written?

Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [reply test/extensions/filters/network/common/redis/client_impl_test.cc]: sure I'll do that

File: source/extensions/filters/network/common/redis/codec_impl.cc [hunk 2]
```diff
@@ -315,8 +315,8 @@ RespValue::CompositeArray::CompositeArrayConstIterator::operator++() {
   return *this;
 }
 
-bool RespValue::CompositeArray::CompositeArrayConstIterator::operator!=(
```
Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [inline source/extensions/filters/network/common/redis/codec_impl.cc]: linter caught this file

File: source/extensions/filters/network/common/redis/redis_command_stats.cc [hunk 3]
```diff
@@ -87,7 +87,7 @@ void RedisCommandStats::updateStats(Stats::Scope& scope, Stats::StatName command
   if (success) {
     counter(scope, {prefix_, command, success_}).inc();
   } else {
-    counter(scope, {prefix_, command, success_}).inc();
```
Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [inline source/extensions/filters/network/common/redis/redis_command_stats.cc]: tests exposed a stats bug

File: test/mocks/event/mocks.cc [hunk 4]
```diff
@@ -10,12 +10,14 @@ using testing::NiceMock;
...
 using testing::ReturnNew;
 using testing::ReturnPointee;
+using testing::ReturnRef;
 using testing::SaveArg;
 
 namespace Envoy {
 namespace Event {
 
 MockDispatcher::MockDispatcher() {
+  ON_CALL(*this, timeSource()).WillByDefault(ReturnRef(time_system_));
```
Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [inline test/mocks/event/mocks.cc]: I've added the ability to use a different time source here- if there is a preferred way to do this please let me know.

**Comment by mattklein123** [comment]
Check format?

/wait

**Comment by FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg** [comment]
Sorry, apparently clang-format was out of date- I should have checked that when I realized I needed to upgrade to bazel v2.

**Comment by mattklein123** [comment]
@FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg is this still WIP or it is it ready for review? If ready for review can you update the PR title?

**Comment by FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg** [comment]
Nope still in progress: Coverage still fails.

**Comment by FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg** [comment]
Ok, ready for review.

File: source/extensions/filters/network/common/redis/redis_command_stats.cc [hunk 5]
```diff
@@ -15,7 +15,7 @@ RedisCommandStats::RedisCommandStats(Stats::SymbolTable& symbol_table, const std
       prefix_(stat_name_set_->add(prefix)),
       upstream_rq_time_(stat_name_set_->add("upstream_rq_time")),
       latency_(stat_name_set_->add("latency")), total_(stat_name_set_->add("total")),
-      success_(stat_name_set_->add("success")), error_(stat_name_set_->add("error")),
+      success_(stat_name_set_->add("success")), failure_(stat_name_set_->add("failure")),
```
Comment (mattklein123) [inline source/extensions/filters/network/common/redis/redis_command_stats.cc]: Does this need a doc update? Or does this now match the docs?

File: test/extensions/filters/network/common/redis/client_impl_test.cc [hunk 6]
```diff
@@ -32,6 +35,13 @@ namespace Common {
 namespace Redis {
 namespace Client {
 
+class MockDispatcherWithTimeOverride : public Event::MockDispatcher {
```
Comment (mattklein123) [inline test/extensions/filters/network/common/redis/client_impl_test.cc]: I believe that you can just derive your test from `TestUsingSimulatedTime` (search for other examples) and avoid doing this. Simulated time will automatically be injected into your test.

File: source/extensions/filters/network/common/redis/redis_command_stats.cc (hunk 5 above)
Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [reply source/extensions/filters/network/common/redis/redis_command_stats.cc]: Good catch; doc has been updated

File: test/extensions/filters/network/common/redis/client_impl_test.cc (hunk 6 above)
Comment (FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg) [reply test/extensions/filters/network/common/redis/client_impl_test.cc]: Excellent, I had not seen that used before- that makes this sort of test much easier!

**Pull request merged**